src/backend/distributed/commands/index.c

```diff
@@ -590,13 +605,30 @@ PreprocessReindexStmt(Node *node, const char *reindexCommand,
 		Oid relationId = ReindexStmtFindRelationOid(reindexStatement, false);
 		MemoryContext relationContext = NULL;
 		Relation relation = NULL;
+		LOCKMODE lockmode = IsReindexWithParam_compat(reindexStatement, "concurrently") ?
+							ShareUpdateExclusiveLock : AccessExclusiveLock;
+
 		if (reindexStatement->kind == REINDEX_OBJECT_INDEX)
 		{
+			/*
+			 * Acquire global lock to prevent concurrent writes or DDL. However,
+			 * we do not bother for REINDEX CONCURRENTLY, since we'll have
+			 * to release the lock.
+			 */
+			if (!IsReindexWithParam_compat(reindexStatement, "concurrently"))
+			{
+				AcquireDistributedLockOnRelations(list_make1(reindexStatement->relation),
+												  lockmode, 0);
+			}
+
 			Oid indOid = RangeVarGetRelid(reindexStatement->relation, NoLock, 0);
 			relation = index_open(indOid, NoLock);
 		}
 		else
 		{
```

Comment:
/REINDEX command accepts INDEX, TABLE, DATABASE, SCHEMA and SYSTEM.

Now, with this else statement, we start to acquire locks for other subcommands other than REINDEX. Is that part of the goal? The PR description only mentions `REINDEX INDEX` subcommand?

Reply:
As of today, Citus only supports `REINDEX INDEX` and `REINDEX TABLE` commands. I tried to add support for both. Let me also add some more comments to make that clear here.